### ISSUE TITLE
Want to ask question about fake chunk in fastbins

### ISSUE THREAD

**Issue by d4em0n** [shellphish/how2heap#71]
Hello, i want to insert fake chunk in fastbins, fake chunk will point into abitrary address. for example, i want to overwrite malloc_hook, and i found fake chunk before malloc_hook where the size is 0x7f. i insert that fake chunk into fastbins and then i malloc, but malloc gives me an error. but when i change the size value in fake chunk with 0x81, malloc success.

is 0x7f is invalid size for fastbins ?, i read article on internet they use fake chunk where the size 0x7f and malloc success return an address of fake_chunk.

i made sample code to create fake chunk.
``` c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char const* argv[])
{
	char data[0x80];
	long *a, *b, *c;
	a = malloc(0x70);
	b = malloc(0x70);
	c = malloc(0x70);
	free(a);
	free(b);
	*(long*)&data[8] = 0x7f;
	printf("data = %p\n", &data);
	*b = (long)&data;
	printf("malloc(0x70) = %p\n", malloc(0x70));
	printf("malloc(0x70) = %p\n", malloc(0x70));
	return 0;
}
```
above script gives me an error.
``` bash
vagrant@vagrant-ubuntu-trusty-64:/vagrant/tmp$ ./fake_chunk 
data = 0x7ffde5eb6c10
malloc(0x70) = 0xa10090
*** Error in `./fake_chunk': malloc(): memory corruption (fast): 0x00007ffde5eb6c20 ***
Aborted (core dumped)
```

when i change the size of fake chunk with 0x81, it's works
``` c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char const* argv[])
{
	char data[0x80];
	long *a, *b, *c;
	a = malloc(0x70);
	b = malloc(0x70);
	c = malloc(0x70);
	free(a);
	free(b);
	*(long*)&data[8] = 0x81;
	printf("data = %p\n", &data);
	*b = (long)&data;
	printf("malloc(0x70) = %p\n", malloc(0x70));
	printf("malloc(0x70) = %p\n", malloc(0x70));
	return 0;
}
```
``` bash
vagrant@vagrant-ubuntu-trusty-64:/vagrant/tmp$ ./fake_chunk 
data = 0x7ffeb847a220
malloc(0x70) = 0x1a73090
malloc(0x70) = 0x7ffeb847a230
```
sorry for my bad english, hope you understand what i mean. thanks for your help

**Comment by m1ghtym0** [comment]
Hi @d4em0n,
There the error you see results from this check in the libc code:
```
              size_t victim_idx = fastbin_index (chunksize (victim));
              if (__builtin_expect (victim_idx != idx, 0))
                  malloc_printerr ("malloc(): memory corruption (fast)");
```
It means the fastbin index for the size you want to allocated `idx` is not equal to the index of the chunk in the fastbin list `victim_index`.
The fastbin index is calculated as:
```
#define fastbin_index(sz) \
((((unsigned int) (sz)) >> (SIZE_SZ == 8 ? 4 : 3)) - 2)
```
Note that sz is calculated with request2size first:
```
#define request2size(req)                                         \
(((req) + SIZE_SZ + MALLOC_ALIGN_MASK < MINSIZE)  ?             \
MINSIZE :                                                      \
((req) + SIZE_SZ + MALLOC_ALIGN_MASK) & ~MALLOC_ALIGN_MASK)
```
We can simplify this to:
`sz = (req + 8 + (16 - 1)) & ~(16-1) = 0x80`

If we insert this into `fastbin_index`, we get:
`idx = (0x80 >> 4) - 2 = 6`

However because you wrote 0x7f into the size field, the ` victim_idx` is calulated as:
`victim_idx = (0x70 >> 4) -2 = 5`

And consequently, `idx != victim_idx`

In other words you forgot to account for the additional memory reserved in `request2size`.
If you change you code to allocated 0x60 instead of 0x70 it should work:-)

**Comment by d4em0n** [comment]
Thanks for your answer @m1ghtym0. I did not notice about `request2size` function.
Terima kasih 🙏
